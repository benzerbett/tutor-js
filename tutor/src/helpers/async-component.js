import { asyncComponent as asyncLoader } from 'react-async-component';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import LoadingComponent from '../components/loading-screen';

export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
