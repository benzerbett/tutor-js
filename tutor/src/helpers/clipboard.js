import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
  isSupported: _.constant(!!SUPPORTED),

  copy() {
    if (!SUPPORTED) { return; }

    try {
      return document.execCommand('copy');
    } catch (e) {
      return console.warn('clipboard copy failed', e);
    }
  },
};
