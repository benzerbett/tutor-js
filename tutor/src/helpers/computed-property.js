import { forIn } from 'lodash';
import { computed } from 'mobx';

export function addComputedProperty(obj, name, fn) {

  Object.defineProperty(obj, name, {
    configurable: true,
    get() {
      const getter = computed(fn, { context: this });
      Object.defineProperty(this, name, {
        configurable: false,
        get: () => getter.get(),
      });
      return getter.get();
    },
  });

}

export function extendHasMany(properties) {
  return (collection) => {
    forIn(properties, (fn, prop) => {
      addComputedProperty(collection, prop, fn);
    });
  };
}
