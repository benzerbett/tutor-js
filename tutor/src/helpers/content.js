import _ from 'underscore';


export default {
  // Convert each number to base 10 with it's position based on index.
  // If section is not present, 0 is set for it
  //   1 becomes 100, 1.1 becomes 101, 3.2 to 302, 3.2.1 -> 30201
  // Useful for sorting
  chapterSectionToNumber(chapter_section) {
    if (_.isString(chapter_section)) { chapter_section = chapter_section.split('.'); }
    if (chapter_section.length === 1) { chapter_section.push(0); } // add a section 0 if it has only a chapter
    let position = -1;
    return _.reduceRight(chapter_section, (memo, num) => memo + (num * Math.pow(100, (position += 1)))
      , 0);
  },
};
