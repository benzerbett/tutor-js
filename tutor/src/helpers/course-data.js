import Courses from '../models/courses-map';

import Router from '../helpers/router';
import TimeHelper from '../helpers/time';

export default {
  getCourseDataProps(courseId) {
    let dataProps;
    if (courseId == null) {
      ((((((({ courseId } = Router.currentParams())))))));
    }
    const course = Courses.get(courseId);
    return dataProps = {
      'data-title': course.name,
      'data-book-title': course.bookName || '',
      'data-appearance': course.appearance_code,
    };
  },

  getCourseBounds(courseId) {
    const course = Courses.get(courseId);

    const start = TimeHelper.getMomentPreserveDate(course.starts_at, TimeHelper.ISO_DATE_FORMAT);
    const end = TimeHelper.getMomentPreserveDate(course.ends_at, TimeHelper.ISO_DATE_FORMAT);

    return { start, end };
  },
};
