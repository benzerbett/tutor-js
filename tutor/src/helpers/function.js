export default {
  // coffeelint: disable=no_empty_functions
  empty() {},
};
// coffeelint: enable=no_empty_functions
