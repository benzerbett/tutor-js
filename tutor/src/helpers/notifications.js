import { NotificationActions } from 'shared';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    // Require router here because requiring it in global scope
    // causes hot module reloading to fail in dev mode because this file is required outside the HMR scope
    const Router = require('./router');
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
