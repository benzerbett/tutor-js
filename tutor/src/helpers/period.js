import { filter, isEmpty, sortBy, toNumber, isNumber } from 'lodash';
import { addComputedProperty } from './computed-property';
import S from './string';

// Used to filter periods by helper methods
const isArchivedCheckFn = period => period.is_archived;
const isActiveCheckFn   = period => !period.is_archived;

const NumberLike = /[^0-9]+/;


const PeriodHelper = {
  getOrder(period) {
    return S.getNumberAndStringOrder(period.name);
  },

  sort(periods) {
    // expects either numbers, names with numbers or just names
    periods = sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const name = period.name.match(/[^0-9]+/ig);
        if (name) {
          return name;
        }
      }
    });
    return sortBy(periods, (period) => { // eslint-disable-line consistent-return
      if (!isNumber(period.name)) {
        const number = period.name.match(/[0-9.-]+/g);
        if (number) {
          return parseFloat(number);
        }
      } else {
        return period.name;
      }
    });
  },

};


export default PeriodHelper;
