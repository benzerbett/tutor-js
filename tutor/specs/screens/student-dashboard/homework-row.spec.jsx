import { React, TimeMock } from '../../helpers';
import Factory from '../../factories';
import HomeworkRow from '../../../src/screens/student-dashboard/homework-row';

describe('Homework Row', function() {
  let props;
  const now = new Date('2017-10-14T12:00:00.000Z');
  TimeMock.setTo(now);

  beforeEach(function() {
    props = {
      event: Factory.studentDashboardTask({ type: 'homework' }),
      course: Factory.course(),
    };
  });

  fit('renders with completed count', function() {
    props.event.correct_exercise_count = null;
    expect.snapshot(<HomeworkRow {...props} />).toMatchSnapshot();
  });

  it('renders complete', function() {
    props.event.complete = true;
    const row = mount(<ReadingRow {...props} />);
    expect(row.find('Col[className="feedback"]').text()).toEqual('Complete');
    row.unmount();
  });

  it('renders close to being due', () => {
    props.event.complete_exercise_count = 1;
    props.event.due_at = new Date(now.getTime() + ( 24*60*60*1000 ));
    props.event.complete = false;
    const row = mount(<ReadingRow {...props} />);
    expect(row.find('Col[className="feedback"]').text()).toEqual('In progress');
    expect(row).toHaveRendered('TutorIcon[type="exclamation-circle"]');
    row.unmount();
  });

});
