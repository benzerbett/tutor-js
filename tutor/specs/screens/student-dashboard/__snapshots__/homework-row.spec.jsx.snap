// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homework Row renders with completed count 1`] = `
<a
  aria-label="Work on homework: turn-key leverage content"
  className="task row homework workable"
  data-event-id={1}
  href="/course/1/task/1"
  onClick={[Function]}
  tabIndex={0}
>
  <div
    className="column-icon col-sm-1 col-2"
  >
    <i
      aria-label="homework icon"
      className="icon icon-lg icon-homework"
    />
  </div>
  <div
    className="title col-sm-6 col-10"
  >
    turn-key leverage content
  </div>
  <div
    className="feedback col-sm-3 col-5"
  >
    <span>
      <span
        className="tour-anchor"
        data-tour-anchor-id="assignment-progress-status"
      >
        2/3 answered
      </span>
    </span>
  </div>
  <div
    className="due-at col-sm-2 col-5"
  >
    <time>
      Oct 11, 7:00am
    </time>
  </div>
</a>
`;
